lib/core/src/Cardano/Pool/Ranking.hs

```diff
@@ -0,0 +1,206 @@
+-- +------------+-------------+------------------------------------------------+
+-- | σ          | @ [0, 1] @  | total relative stake of the pool               |
+-- +------------+-------------+------------------------------------------------+
+module Cardano.Pool.Ranking
+    (
+      -- * Formulas
+      desirability
+
+    , saturatedPoolRewards
+    , saturatedPoolSize
+
+      -- * Types
+    , EpochConstants (..)
+    , Pool (..)
+    , RelativeStakeOf (..)
+    , mkRelativeStake
+    , Lovelace (..)
+    , Margin
+    , unsafeMkMargin
+    , getMargin
+    , NonNegative (..)
+    , Positive (..)
+    , unsafeToPositive
+    , unsafeToNonNegative
+    )
+    where
+
+import Prelude
+
+import GHC.Generics
+    ( Generic )
+import GHC.TypeLits
+    ( Symbol )
+
+--------------------------------------------------------------------------------
+-- Formulas from spec
+--------------------------------------------------------------------------------
+
+-- | Non-myopic pool desirability according to section 5.6.1.
+--
+-- Is /not/ affected by oversaturation nor pool stake in general.
+desirability
+    :: EpochConstants
+    -> Pool
+    -> Double
+desirability constants pool
+    | f_saturated <= c = 0
+    | otherwise    = (f_saturated - c) * (1 - m)
+  where
+    f_saturated = saturatedPoolRewards constants pool
+    m = getMargin $ margin pool
+    c = getNonNegative $ getLovelace $ cost pool
+
+-- | Total rewards for a pool if it were saturated.
+--
+-- When a0 = 0 this reduces to just p*R*z0 (tested
+-- by @prop_saturatedPoolRewardsReduces@)
+saturatedPoolRewards :: EpochConstants -> Pool -> Double
+saturatedPoolRewards constants pool =
+    let
+        a0 = getNonNegative $ leaderStakeInfluence constants
+        z0 = unRelativeStake $ saturatedPoolSize constants
+        s = unRelativeStake $ leaderStake pool
+        _R = getNonNegative $ getLovelace $ totalRewards constants
+        p = getNonNegative $ recentAvgPerformance pool
+        -- ^ technically \hat{p} in the spec
+    in
+        (p * _R) / (1 + a0)
+        * (z0 + ((min s z0) * a0))
+
+-- | Determines z0, i.e 1 / k
+saturatedPoolSize :: EpochConstants -> RelativeStakeOf "pool"
+saturatedPoolSize constants =
+    RelativeStake $ 1 / fromIntegral (getPositive $ optimalNumberOfPools constants)
+
+--------------------------------------------------------------------------------
+-- Types
+--------------------------------------------------------------------------------
+
+data EpochConstants = EpochConstants
+    { leaderStakeInfluence :: NonNegative Double
+      -- ^ a_0
+    , optimalNumberOfPools :: Positive Int
+      -- ^ k
+    , totalRewards :: Lovelace
+      -- ^ Total rewards in an epoch. "R" in the spec.
+    } deriving (Show, Eq, Generic)
+
+data Pool = Pool
+    { leaderStake :: RelativeStakeOf "pool leader"
+    , cost :: Lovelace
+    , margin :: Margin
+    , recentAvgPerformance :: NonNegative Double
```

Comment:
`NonNegative Double` would suppose that the performance could be greater than 1, which isn't the case. Incidentally `margin`, `leaderStake`  and `recentAvgPerformance` all have the same unit / type which should be equivalent to a fractional number between 0 and 1.

Reply:
>  performance could be greater than 1, which isn't the case

I cannot conclude this from the spec.

> Even when a pool’s true performance is 1, its actual rewards might be less than its optimal rewards, because of the randomness in re. Likewise, sometimes a pool’s actual rewards will be **higher than their optimal rewards**, if they are **lucky** in the leader election process.
>
> **These effects will balance each other**, so that a well performing pool will get their optimal rewards on average over multiple epochs.

and from the formulas I don't see how `\bar{p}` would be constrained to `[0,1]`:

<img width="801" alt="Skärmavbild 2020-01-17 kl  11 48 04" src="https://user-images.githubusercontent.com/304423/72606639-3d186600-391f-11ea-834d-2ca2c5d97213.png">

I have another concern about clamping the **pre**-averaged performances, especially, with `min 1` , but that is separate to this.

Reply:
If there is a choice to be made whether to restrict to `[0,1]` or not it seems better to me to not make that choice here. Especially if the type of `recentAvgPerformance` is going to error when a value is out of bounds.

Reply:
Fair enough :+1: